Purchases/Purchasing/StoreKitAbstractions/PromotionalOffer.swift

```diff
@@ -33,6 +33,17 @@ extension PromotionalOffer {
                                  signature: self.signature,
                                  timestamp: self.timestamp as NSNumber)
     }
+
+    @available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+    var sk2PurchaseOption: Product.PurchaseOption {
+        return .promotionalOffer(
+            offerID: self.identifier,
+            keyID: self.keyIdentifier,
+            nonce: self.nonce,
+            signature: self.signature.data(using: .utf8)!,
```

Comment:
this comes from a backend request, right? 

If it does, then I think if there's an issue decoding, we should fail elegantly (perhaps we can throw here and propagate in methods that call it), but not hard crash

Reply:
This would only return `nil` if `signature` can't be represented in `utf8`. I don't think our backend would send that?
Maybe instead of having to handle such an unlikely error, we could use
```
self.signature.data(using: .utf8, allowLossyConversion: true)
```
> An `NSData` object containing a representation of the receiver encoded using encoding. Returns nil if flag is NO and the receiver can’t be converted without losing some information (such as accents or case).
> If flag is YES and the receiver can’t be converted without losing some information, some characters may be removed or altered in conversion. For example, in converting a character from NSUnicodeStringEncoding to NSASCIIStringEncoding, the character ‘Á’ becomes ‘A’, losing the accent.

I think that's the best compromise, what do you think?

Reply:
If we allow for lossy conversion, we'll have a non-working signature, right? So it will just not work? 

I think we should always be defensive and consider the possibility of a backend bug. If the backend messes up the signature data for whatever reason, I don't think we should have thousands of apps crashing until the backend recovers, when instead we could safely just disallow promo offers until the backend comes back.

We can handle that with the lossy conversion and having that implicitly not work, but I'd vote for being explicit, and if it can't be encoded, log an error that clearly states that the backend gave us weird info, then return either empty or an error (whatever's more appropriate for the method).

Reply:
You're totally right. I'm all for detecting errors early, and in that way users could notify us if the server is sending any bad data. I don't know what I was thinking.
I just changed this to throw `ErrroCode.unexpectedBackendResponse`.